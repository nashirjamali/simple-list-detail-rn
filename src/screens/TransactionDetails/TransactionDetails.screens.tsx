import React, {createContext, useContext, useState} from 'react';
import {Alert, View} from 'react-native';
import {useNavigation} from '@react-navigation/native';
import {StackNavigationProp} from '@react-navigation/stack';
import Clipboard from '@react-native-clipboard/clipboard';

import type {
  RootStackParamList,
  Transaction,
  TransactionContextType,
} from '../../@types';
import {
  ArrowRightIcon,
  Button,
  CopyIcon,
  GeneralText,
  Layout,
} from '../../components';
import {BUTTON_VARIANTS, COLORS, GENERAL_TEXT_VARIANTS} from '../../constants';
import {formatDateToIndonesian, formatToIDR} from '../../utils';

import styles from './TransactionDetails.styles';

const TransactionContext = createContext<TransactionContextType | undefined>(
  undefined,
);

/**
 * Provider component for managing transaction state.
 *
 * @param {React.ReactNode} children - The child components to be rendered within the provider.
 * @returns {JSX.Element} The rendered TransactionProvider component.
 */
export const TransactionProvider: React.FC<{children: React.ReactNode}> = ({
  children,
}) => {
  const [transaction, setTransaction] = useState<Transaction | null>(null);

  return (
    <TransactionContext.Provider value={{transaction, setTransaction}}>
      {children}
    </TransactionContext.Provider>
  );
};

/**
 * Custom hook to access the transaction context.
 *
 * @returns {TransactionContextType} The current transaction context.
 * @throws {Error} If used outside of a TransactionProvider.
 */
export const useTransactionContext = () => {
  const context = useContext(TransactionContext);
  if (!context) {
    throw new Error(
      'useTransactionContext must be used within a TransactionProvider',
    );
  }
  return context;
};

/**
 * TransactionDetail component displays the details of a specific transaction.
 *
 * @returns {JSX.Element} The rendered TransactionDetail component.
 */
const TransactionDetail: React.FC = () => {
  const {transaction} = useTransactionContext();
  const navigation = useNavigation<StackNavigationProp<RootStackParamList>>();

  const handleCloseButton = () => {
    navigation.goBack();
  };

  const handleCopy = (id: string) => {
    Clipboard.setString(id);
    Alert.alert('Copied to Clipboard');
  };

  return (
    <Layout>
      {transaction && (
        <View>
          <View style={styles.transactionCode}>
            <GeneralText variant={GENERAL_TEXT_VARIANTS.HEADER}>
              ID TRANSAKSI: #{transaction.id}
            </GeneralText>
            <Button
              title="Copy"
              icon={
                <CopyIcon
                  color={COLORS.BRAND}
                  width={20}
                  height={20}
                  onPress={() => handleCopy(transaction.id)}
                />
              }
              variant={BUTTON_VARIANTS.ICON}
            />
          </View>
          <View style={styles.header}>
            <GeneralText variant={GENERAL_TEXT_VARIANTS.HEADER}>
              DETAIL TRANSAKSI
            </GeneralText>
            <Button
              title="Tutup"
              variant={BUTTON_VARIANTS.NORMAL}
              onPress={handleCloseButton}
            />
          </View>

          <View style={styles.detailContainer}>
            <View style={styles.detailItem}>
              <GeneralText variant={GENERAL_TEXT_VARIANTS.HEADER}>
                {transaction.sender_bank.toUpperCase()}
              </GeneralText>
              <ArrowRightIcon />
              <GeneralText variant={GENERAL_TEXT_VARIANTS.HEADER}>
                {transaction.beneficiary_bank.toUpperCase()}
              </GeneralText>
            </View>

            <View style={styles.detailItem}>
              <View style={{flex: 1}}>
                <GeneralText variant={GENERAL_TEXT_VARIANTS.HEADER}>
                  {transaction.beneficiary_name}
                </GeneralText>
                <GeneralText variant={GENERAL_TEXT_VARIANTS.NORMAL}>
                  {transaction.account_number}
                </GeneralText>
              </View>
              <View style={{flex: 1}}>
                <GeneralText variant={GENERAL_TEXT_VARIANTS.HEADER}>
                  NOMINAL
                </GeneralText>
                <GeneralText variant={GENERAL_TEXT_VARIANTS.NORMAL}>
                  {formatToIDR(transaction.amount)}
                </GeneralText>
              </View>
            </View>

            <View style={styles.detailItem}>
              <View style={{flex: 1}}>
                <GeneralText variant={GENERAL_TEXT_VARIANTS.HEADER}>
                  BERITA TRANSFER
                </GeneralText>
                <GeneralText variant={GENERAL_TEXT_VARIANTS.NORMAL}>
                  {transaction.remark}
                </GeneralText>
              </View>
              <View style={{flex: 1}}>
                <GeneralText variant={GENERAL_TEXT_VARIANTS.HEADER}>
                  KODE UNIK
                </GeneralText>
                <GeneralText variant={GENERAL_TEXT_VARIANTS.NORMAL}>
                  {transaction.unique_code}
                </GeneralText>
              </View>
            </View>

            <View style={styles.detailItem}>
              <View style={{flex: 1}}>
                <GeneralText variant={GENERAL_TEXT_VARIANTS.HEADER}>
                  WAKTU DIBUAT
                </GeneralText>
                <GeneralText variant={GENERAL_TEXT_VARIANTS.NORMAL}>
                  {formatDateToIndonesian(transaction.created_at)}
                </GeneralText>
              </View>
            </View>
          </View>
        </View>
      )}
    </Layout>
  );
};

export default TransactionDetail;
