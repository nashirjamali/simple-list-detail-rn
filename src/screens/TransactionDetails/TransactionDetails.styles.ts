import {StyleSheet} from 'react-native';

import type {TransactionDetailStyles} from '../../@types';
import {COLORS} from '../../constants';

const styles: TransactionDetailStyles = StyleSheet.create({
  transactionCode: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    gap: 5,
    padding: 10,
    backgroundColor: COLORS.WHITE,
    borderBottomColor: COLORS.GRAY,
    borderBottomWidth: 1,
  },
  header: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    padding: 10,
    backgroundColor: COLORS.WHITE,
    borderBottomColor: COLORS.GRAY,
    borderBottomWidth: 1,
  },
  detailContainer: {
    padding: 10,
    display: 'flex',
    gap: 10,
  },
  detailItem: {
    display: 'flex',
    flexDirection: 'row',
    gap: 5,
    alignItems: 'center',
  },
});

export default styles;
