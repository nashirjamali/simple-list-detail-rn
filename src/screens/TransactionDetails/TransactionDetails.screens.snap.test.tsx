import React, {type ReactNode} from 'react';
import {render} from '@testing-library/react-native';

import type {Transaction} from '../../@types';

import {
  TransactionContext,
  TransactionProvider,
} from './TransactionDetails.screens';
import TransactionDetails from './TransactionDetails.screens';
import {NavigationContainer} from '@react-navigation/native';

const mockTransaction: Transaction = {
  id: '12345',
  sender_bank: 'Bank A',
  beneficiary_bank: 'Bank B',
  beneficiary_name: 'John Doe',
  account_number: '1234567890',
  amount: 1000000,
  remark: 'Payment for services',
  unique_code: 50,
  created_at: '2024-01-01T12:00:00Z',
  completed_at: '2024-01-01T12:00:00Z',
  fee: 10000,
  status: 'SUCCESS',
};

const MockTransactionProvider: React.FC<{
  transaction: Transaction | null;
  children: ReactNode;
}> = ({transaction, children}) => {
  return (
    <NavigationContainer>
      <TransactionContext.Provider
        value={{transaction, setTransaction: jest.fn()}}>
        {children}
      </TransactionContext.Provider>
    </NavigationContainer>
  );
};

describe('TransactionDetail Component', () => {
  it('renders correctly with transaction data', () => {
    const {toJSON} = render(
      <MockTransactionProvider transaction={mockTransaction}>
        <TransactionDetails />
      </MockTransactionProvider>,
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly when no transaction is provided', () => {
    const {toJSON} = render(
      <MockTransactionProvider transaction={null}>
        <TransactionDetails />
      </MockTransactionProvider>,
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
