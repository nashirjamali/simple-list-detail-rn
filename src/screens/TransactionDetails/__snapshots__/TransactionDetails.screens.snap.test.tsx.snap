// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionDetail Component renders correctly when no transaction is provided 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    {
      "backgroundColor": "#F5F9F8",
      "padding": 10,
    }
  }
/>
`;

exports[`TransactionDetail Component renders correctly with transaction data 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    {
      "backgroundColor": "#F5F9F8",
      "padding": 10,
    }
  }
>
  <View>
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderBottomColor": "#999999",
          "borderBottomWidth": 1,
          "display": "flex",
          "flexDirection": "row",
          "gap": 5,
          "padding": 10,
        }
      }
    >
      <Text
        style={
          {
            "color": "#121212",
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        ID TRANSAKSI: #
        12345
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "flexDirection": "row",
            "gap": 2,
            "justifyContent": "space-between",
            "opacity": 1,
          }
        }
      >
        <RNSVGSvgView
          align="xMidYMid"
          bbHeight={20}
          bbWidth={20}
          color="#FB6544"
          fill="none"
          focusable={false}
          height={20}
          meetOrSlice={0}
          minX={0}
          minY={0}
          onPress={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          responsible={true}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              {
                "flex": 0,
                "height": 20,
                "width": 20,
              },
            ]
          }
          vbHeight={24}
          vbWidth={24}
          width={20}
        >
          <RNSVGGroup
            fill={null}
            propList={
              [
                "fill",
              ]
            }
          >
            <RNSVGPath
              d="M6 11c0-2.828 0-4.243.879-5.121C7.757 5 9.172 5 12 5h3c2.828 0 4.243 0 5.121.879C21 6.757 21 8.172 21 11v5c0 2.828 0 4.243-.879 5.121C19.243 22 17.828 22 15 22h-3c-2.828 0-4.243 0-5.121-.879C6 20.243 6 18.828 6 16v-5z"
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              propList={
                [
                  "stroke",
                  "strokeWidth",
                ]
              }
              stroke={
                {
                  "payload": 4294665540,
                  "type": 0,
                }
              }
              strokeWidth={1.5}
            />
            <RNSVGPath
              d="M6 19a3 3 0 01-3-3v-6c0-3.771 0-5.657 1.172-6.828C5.343 2 7.229 2 11 2h4a3 3 0 013 3"
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              propList={
                [
                  "stroke",
                  "strokeWidth",
                ]
              }
              stroke={
                {
                  "payload": 4294665540,
                  "type": 0,
                }
              }
              strokeWidth={1.5}
            />
          </RNSVGGroup>
        </RNSVGSvgView>
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderBottomColor": "#999999",
          "borderBottomWidth": 1,
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "padding": 10,
        }
      }
    >
      <Text
        style={
          {
            "color": "#121212",
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        DETAIL TRANSAKSI
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "flexDirection": "row",
            "gap": 2,
            "justifyContent": "space-between",
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            {
              "color": "#FB6544",
              "fontSize": 14,
            }
          }
        >
          Tutup
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "display": "flex",
          "gap": 10,
          "padding": 10,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "gap": 5,
          }
        }
      >
        <Text
          style={
            {
              "color": "#121212",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          BANK A
        </Text>
        <RNSVGSvgView
          align="xMidYMid"
          bbHeight={20}
          bbWidth={20}
          fill="none"
          focusable={false}
          height={20}
          meetOrSlice={0}
          minX={0}
          minY={0}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              {
                "flex": 0,
                "height": 20,
                "width": 20,
              },
            ]
          }
          vbHeight={24}
          vbWidth={24}
          width={20}
        >
          <RNSVGGroup
            fill={null}
            propList={
              [
                "fill",
              ]
            }
          >
            <RNSVGPath
              d="M13.233 16.457a.75.75 0 101.034 1.086l-1.034-1.086zm6.284-3.914a.75.75 0 10-1.034-1.086l1.034 1.086zm-1.034 0a.75.75 0 101.034-1.086l-1.034 1.086zm-4.216-6.086a.75.75 0 10-1.034 1.086l1.034-1.086zM19 12.75a.75.75 0 000-1.5v1.5zm-14-1.5a.75.75 0 000 1.5v-1.5zm9.267 6.293l5.25-5-1.034-1.086-5.25 5 1.034 1.086zm5.25-6.086l-5.25-5-1.034 1.086 5.25 5 1.034-1.086zM19 11.25H5v1.5h14v-1.5z"
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              propList={
                [
                  "fill",
                ]
              }
            />
          </RNSVGGroup>
        </RNSVGSvgView>
        <Text
          style={
            {
              "color": "#121212",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          BANK B
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "gap": 5,
          }
        }
      >
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <Text
            style={
              {
                "color": "#121212",
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            John Doe
          </Text>
          <Text
            style={
              {
                "color": "#121212",
                "fontSize": 14,
              }
            }
          >
            1234567890
          </Text>
        </View>
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <Text
            style={
              {
                "color": "#121212",
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            NOMINAL
          </Text>
          <Text
            style={
              {
                "color": "#121212",
                "fontSize": 14,
              }
            }
          >
            Rp1.000.000
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "gap": 5,
          }
        }
      >
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <Text
            style={
              {
                "color": "#121212",
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            BERITA TRANSFER
          </Text>
          <Text
            style={
              {
                "color": "#121212",
                "fontSize": 14,
              }
            }
          >
            Payment for services
          </Text>
        </View>
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <Text
            style={
              {
                "color": "#121212",
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            KODE UNIK
          </Text>
          <Text
            style={
              {
                "color": "#121212",
                "fontSize": 14,
              }
            }
          >
            50
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "gap": 5,
          }
        }
      >
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <Text
            style={
              {
                "color": "#121212",
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            WAKTU DIBUAT
          </Text>
          <Text
            style={
              {
                "color": "#121212",
                "fontSize": 14,
              }
            }
          >
            1 Januari 2024
          </Text>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
